go/arrow/memory/internal/cgoalloc/allocator.go

```diff
@@ -0,0 +1,99 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build ccalloc
+
+package cgoalloc
+
+// #cgo !windows pkg-config: arrow
+// #cgo CXXFLAGS: -std=c++14
+// #cgo windows LDFLAGS:  -larrow
+// #include "allocator.h"
+import "C"
+import (
+	"reflect"
+	"unsafe"
+)
+
+// CGOMemPool is an alias to the typedef'd uintptr from the allocator.h file
+type CGOMemPool = C.ArrowMemoryPool
+
+// CgoPoolAlloc allocates a block of memory of length 'size' using the memory
+// pool that is passed in.
+func CgoPoolAlloc(pool CGOMemPool, size int) []byte {
+	var out *C.uint8_t
+	C.arrow_pool_allocate(pool, C.int64_t(size), (**C.uint8_t)(unsafe.Pointer(&out)))
+
+	var ret []byte
+	s := (*reflect.SliceHeader)(unsafe.Pointer(&ret))
+	s.Data = uintptr(unsafe.Pointer(out))
+	s.Len = size
+	s.Cap = size
+
+	return ret
+}
+
+// CgoPoolRealloc calls 'reallocate' on the block of memory passed in which must
+// be a slice that was returned by CgoPoolAlloc or CgoPoolRealloc.
+func CgoPoolRealloc(pool CGOMemPool, size int, b []byte) []byte {
+	oldSize := C.int64_t(len(b))
+	data := (*C.uint8_t)(unsafe.Pointer(&b[0]))
+	C.arrow_pool_reallocate(pool, oldSize, C.int64_t(size), &data)
+
+	var ret []byte
+	s := (*reflect.SliceHeader)(unsafe.Pointer(&ret))
+	s.Data = uintptr(unsafe.Pointer(data))
+	s.Len = size
+	s.Cap = size
+
+	return ret
+}
+
+// CgoPoolFree uses the indicated memory pool to free a block of memory. The
+// slice passed in *must* be a slice which was returned by CgoPoolAlloc or
+// CgoPoolRealloc.
+func CgoPoolFree(pool CGOMemPool, b []byte) {
+	if len(b) == 0 {
```

Comment:
I don't see a corresponding condition in `CgoPoolAlloc` and `CgoPoolRealloc`, won't this leak 0-byte allocated areas?

Reply:
fair point, i assumed that calling allocate with a 0 byte size wouldn't actually allocate anything and thus there would be nothing to leak. i've added a corresponding check to alloc and realloc to address this